File: 1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.cpp
class Solution {
public:
    int constrainedSubsetSum(vector<int>& nums, int k) {
        int res = nums[0];
        deque<pair<int,int>>dq;
        dq.push_back({res,0});
        for(int i = 1;i<nums.size();i++){
            int idx = dq.front().second;
            if(idx < (i-k))dq.pop_front();
            int val = dq.front().first;
            int ans = nums[i];
            if(val > 0) ans += val;
            res = max(res,ans);
            while((!dq.empty()) && ans > dq.back().first){
                dq.pop_back();
            }
          //  cout<<ans<<" "<<i<<endl;
            dq.push_back({ans,i});
        }
        return res;
    }
};

File: 1286-constrained-subsequence-sum/1286-constrained-subsequence-sum.java
class Solution {
    class pair{
        int first,second;
        pair(int a,int b){
            first = a;
            second = b;
        }
    }
    public int constrainedSubsetSum(int[] nums, int k) {
        int res = nums[0];
        Deque<pair>dq = new ArrayDeque<>();
        pair temp = new pair(res,0);
        dq.addLast(temp);
        for(int i = 1;i<nums.length;i++){
            int idx = dq.getFirst().second;
            if(idx < (i-k)){
                pair p1 = dq.removeFirst();
            }
            int val = dq.getFirst().first;
            int ans = nums[i];
            if(val > 0) ans += val;
            res = Math.max(res,ans);
            
            while((!dq.isEmpty()) && ans > dq.getLast().first){
              temp = dq.removeLast();
            }

            dq.addLast(new pair(ans,i));
        }
        return res;
    }
}